### ISSUE TITLE
Account.sign() fails 343 out of 25600 times

### ISSUE THREAD

**Issue by miohtama** [ethereum/web3.py#466]
* Version: master
* Python: 3.5
* OS: osx

### What was wrong?

I am trying to use Account.sign() to sign orders against 0x Exchange.sol contract. Account.sign() has been grabbed from the master for this purpose. Account.sign() seems to be produce a signature that Exchange.sol does not accept 343 out of 25600 times. Account.sign() input is seemingly random (SHA3 hash of 0x order parameters).

The private key used to sign is 0x5e95384d8050109aab08c1922d3c230739bc16976553c317e5d0b87b59371f2a and does not change between samples.

I am using Populus 1.11.x testrpc for the EVM bit.

Here is the full log of 343 failed hashes (to detect a pattern): https://gist.github.com/miohtama/93b87a963dcc7c944b5d5e6142815347

Below is my test case

```
def test_order_signature(web3, exchange, parties_ready_to_trade, bought_token, sold_token, fee_recipient, private_key_hex, private_key_account):
    """Python and Solidity agree how to sign orders."""

    buyer, seller = parties_ready_to_trade

    failure_count = 0
    for i in range(0, 25600):
        to_buy = 300
        to_sell = 200
        maker_fee = random.randint(0, 100)
        taker_fee = random.randint(0, 100)

        fee_recipient = "0x0000000000000000000000000000000000000000"

        o = create_test_order(
            web3,
            exchange_address=exchange.address,
            fee_recipient=fee_recipient,
            maker_address=seller,
            taker_address=buyer,
            maker_token=sold_token.address,
            taker_token=bought_token.address,
            maker_token_amount=to_sell,
            taker_token_amount=to_buy,
            maker_fee=maker_fee,
            taker_fee=taker_fee,
        )

        # Solidity takes our signature as granted
        hash = o.calculate_hash()
        hash_bytes = decode_hex(hash)

        data = o.sign_private_key(private_key_hex)

# is same as
#        hash = self.calculate_hash()
#        hash_bytes = decode_hex(hash)
#
  #      a = Account()
 #       sign_data = a.sign(hash_bytes, private_key=decode_hex(private_key_hex))
   #     signature_params = {
     #       "hash": hash,
      #      "hash_bytes": hash_bytes,
       #     "v": sign_data["v"],
        #    "r": sign_data["r"],
         #   "s": sign_data["s"],
        #}
        #return signature_params


        try:
            assert exchange.call().isValidSignature(seller, hash_bytes,
                                                data["v"],
                                                data["r"],
                                                data["s"],)
        except:
            print("Failed signature is, number:", i, "data:", data)
            failure_count += 1

    print("Failure count is", failure_count)
```

How Exchange.sol validates the signed message: https://github.com/0xProject/contracts/blob/master/contracts/Exchange.sol#L454

### How can it be fixed?

Oh boy if I knew I would tell you.

**Comment by miohtama** [comment]
Looks like when `sign_message_hash` gets a signature where the first byte is `00` it fails:

```

**Comment by miohtama** [comment]
Here is my naive fix

```
def sign_message_hash(key, msg_hash):
    signature = key.sign_msg_hash(msg_hash)
    (v_standard, r, s) = signature.vrs
    v = v_standard + V_OFFSET

    # Pad it like Solidity expects for bytes32
    r_b = to_bytes(r)
    s_b = to_bytes(s)
    r_b = pad_left(r_b, 32, b"\0")
    s_b = pad_left(s_b, 32, b"\0")

    eth_signature_bytes = b''.join(map(to_bytes, (r_b, s_b, v)))

    return (v, r_b, s_b, eth_signature_bytes)

```

**Comment by carver** [comment]
While investigating this issue in signed messages, I also looked into transactions. What I found is that web3js v1 0-pads `r` and `s` values in signed transactions. But things started to get hairy when implementing that and I got suspicious. So I looked at geth and parity. Neither of them 0-pad their r and s values in the raw transaction.

Interestingly, that means that exactly the same transaction is created with different hashes in geth/parity vs in web3js. I imagine that clients will import with or without leading 0s just fine. But perhaps this is a surprise to some.

However, signed messages in web3.js, geth, and the `eth_keys` library all produce zero-padded 65-byte signatures. That seems like an obvious thing to fix.

So I believe the conclusion is this:

- Zero-pad the `r` and `s` values when represented as hex inside a combined "signature bytes" value
- Continue to *not* zero-pad `r` and `s` values in transactions (consistent with geth/parity, inconsistent with the web3.js)
- When returned as individual values, `r` and `s` should be returned as integers instead of bytes, to avoid confusion and better match the underlying representation
